### PR TITLE
MG

### PR DESCRIPTION
**Pull request by DanielRichtmann**

### CODE REVIEW

**Comment by lehner** [comment]
Hi Daniel,

two comments:

- can you change interface from
  `t = g.timer(["setup", "prec", "mat", "ortho", "linalg", "total"])`
to
`t = g.timer("setup", "prec", "mat", "ortho", "linalg", "total")`
by using 

 `class timer:
    def __init__(self, *names):
        self.dt = {key: 0.0 for key in names}
`

- in matrix_operators you added __call__(first,second,third,fourth),
  can you change this to __call__(dst,src,*extra) so that we still have
  a clear convention from dst,src field and are future-proof for extra
 arguments?

Thanks!

-- C

**Comment by DanielRichtmann** [comment]
Commit https://github.com/lehner/gpt/pull/35/commits/80c07fbbeaddb450461d24647200dcc96925519c resolves your first suggestion.

**Comment by DanielRichtmann** [comment]
This also requires some changes in Grid. I opened https://github.com/lehner/Grid/pull/13 for that.

**Comment by DanielRichtmann** [comment]
This should now be in a state ready for a first merge.
CI is passing locally, will see what github says.

**Comment by lehner** [comment]
Very nice, am busy for the next few hours but will try to go through and then merge before the GPT meeting tomorrow.

File: lib/gpt/qcd/fermion/operator.py [hunk 1]
```diff
@@ -196,3 +208,166 @@ def prop(dst_sc, src_sc):
...
+        self.params_constructor = params
+        self.Ls = Ls
+        self.otype = otype
+
+        # derived objects
+        self.A_grid = A[0].grid
+        # self.A_grid_eo = gpt.grid(
+        #     self.A_grid.gdimensions, self.A_grid.precision, gpt.redblack
+        # )
+        if Ls is not None:
```
Comment (aragon999) [inline lib/gpt/qcd/fermion/operator.py:228]: I think this if statement can be removed

**Comment by lehner** [comment]
Will merge this now but have a list of reasonably minor modifications that I want to implement after merging.  Daniel, you deserve a break for a day and maybe you can check my modifications/pull next Monday.

Excellent job, congratulations!

**Pull request merged**